cli/generate/src/lib.rs

```diff
@@ -42,14 +49,104 @@ struct GeneratedParser {
 pub const ALLOC_HEADER: &str = include_str!("templates/alloc.h");
 pub const ARRAY_HEADER: &str = include_str!("templates/array.h");
 
+pub type GenerateResult<T> = Result<T, GenerateError>;
+
+#[derive(Debug, Error, Serialize)]
+pub enum GenerateError {
+    #[error("Error with specified path -- {0}")]
+    GrammarPath(String),
+    #[error("{0}")]
+    IO(String),
+    #[error("{0}")]
+    LoadGrammarFile(#[from] LoadGrammarFileError),
+    #[error("{0}")]
+    ParseGrammar(#[from] ParseGrammarError),
+    #[error("{0}")]
+    GenerateParser(#[from] GenerateParserError),
+}
+
+impl From<std::io::Error> for GenerateError {
+    fn from(value: std::io::Error) -> Self {
+        Self::IO(value.to_string())
+    }
+}
+
+pub type GenerateParserResult<T> = Result<T, GenerateParserError>;
+
+#[derive(Debug, Error, Serialize)]
+pub enum GenerateParserError {
+    #[error("{0}")]
+    Prepare(#[from] PrepareGrammarError),
+    #[error("{0}")]
+    VariableInfo(#[from] VariableInfoError),
+    #[error("{0}")]
+    BuildTables(#[from] ParseTableBuilderError),
+}
```

Comment:
I think you can flatten this into `GenerateError`, and avoid a bit of unnecessary indirection